docs/architecture-decision-records/20220121-conduit-plugin-architecture.md

```diff
@@ -0,0 +1,322 @@
+# Conduit Plugin Architecture
+
+## Summary
+
+This doc describes the design for the conduit plugin architecture. It spans from how Conduit communicates with built-in
+and standalone plugins, where and how the interface is defined, how the plugin SDK is defined.
+
+## Context
+
+We want to provide a good development experience to plugin developers, we also want to be able to ship Conduit with real
+built-in plugins (compiled into the Conduit binary), we want to make it as easy as possible to write plugins in any
+programming language and we want the SDK to be decoupled from Conduit and be able to change without changing Conduit
+itself.
+
+## Decision
+
+The decision can be broken up into 4 parts:
+
+- We defined the plugin interface in gRPC
+- We defined the interface used internally in Conduit to communicate with plugins
+- We introduced plugin registries to support built-in and standalone plugins
+- We defined the interface for the plugin SDK
+
+### Conduit plugin interface (gRPC)
+
+The Conduit plugin interface is defined in gRPC, is standalone (does not depend on Conduit or the SDK) and lives
+in https://github.com/conduitio/conduit-plugin. This repository is the only common thing between Conduit and a plugin,
+meaning that Conduit uses the interface to interact with plugins and plugins implement the interface.
+
+The proto files define the messages and gRPC interface that needs to be implemented by the connector plugin. The
+interface is uploaded to the [buf schema registry](https://buf.build/conduitio/conduit-plugin), which allows us to
+easily generate the code in any programming language. For Go we
+use [remote generation](https://docs.buf.build/bsr/remote-generation/overview) so we don’t have to manually generate any
+code.
+
+Alongside the proto files, the repository contains helper structs and methods that hide the gRPC implementation from the
+user without changing any semantics. No other code needs to know that the communication happens through gRPC (i.e.
+neither Conduit nor the plugin SDK).
+
+There are 3 gRPC services that a plugin needs to implement - `SourcePlugin`, `DestinationPlugin` and `SpecifierPlugin`.
```

Comment:
I didn't explicitly mention it, but the plugin can pass `nil` to [`Serve`](https://github.com/ConduitIO/conduit/pull/84/files#diff-b8b88c00c1bb03a2238d9ad4aace34d8e5586115ace96220afd9984ef09c4b58R25) and the SDK will expose an unimplemented plugin (i.e. all functions return an error).

Reply:
That sounds cool. If you don't mind, can we add a sentence or two about that, so that whoever is implementing a plugin knows that? I think it's useful info that you can simply give a nil (and not, for example, implement dummy functions yourself).

Reply:
I added a sentence, you can find it lower in the document in the part about the plugin SDK.